File: WMC_simulations/run_simulations/run_simulations.m
clear
close all

model_resolution_in_mm = 0;

% Lambdas = 400:10:490;
Lambdas = 400;
run_in_cluster = 0;
nb_repeat = 1;%50; %Nb of repetitions used in MCX
simu_repeat = 3; %Larger number of repeat (avoid Matlab crash due to large txt files)
nb_photons = 10;%1e6;

out_path = 'output/';
in_img_path = '../images/Patient1/';
%in_img_path = '';


% Add path
addpath('../functions');
if run_in_cluster == 1
    addpath('/pbs/home/c/ccaredda/private/mcx/utils');
    addpath('/pbs/home/c/ccaredda/private/mcxlab');
else
    addpath('/home/caredda/Soft/mcx/utils');
    addpath('/home/caredda/Soft/mcxlab')
    % addpath('/home/admin/Software/mcx-linux-x86_64-v2023/mcx/utils');
    % addpath('/home/admin/Software/mcxlab-linux-x86_64-v2023/mcxlab');  
end



% Create output directory
if ~ isfolder(out_path)
    mkdir(out_path);
end


%Process model info
info_model = process_model_info(nb_photons,nb_repeat,in_img_path,model_resolution_in_mm);

save(strcat(out_path,'cst.mat'),'info_model');
f = fopen(strcat(out_path,'cst.txt'),'w');
fprintf(f,'%s %d\n','nb_photons ',nb_photons);
fprintf(f,'%s %d\n','repetitions ',nb_repeat*simu_repeat);
fprintf(f,'%s %f\n','unitinmm ',info_model.cfg.unitinmm);
fprintf(f,'%s %d\n','vol_rows ',size(info_model.cfg.vol,1));
fprintf(f,'%s %d\n','vol_cols ',size(info_model.cfg.vol,2));
fclose(f);

%Process optical properties
Optical_prop = process_optical_properties(Lambdas);

% process simulations
for l=1:length(Lambdas)
    disp(strcat("Simulation lambda ",num2str(Lambdas(l))))
    
    for s=1:simu_repeat
        % %  % detector output
        % % output_det.ppath = detphoton.ppath; % cummulative path lengths in each medium (partial pathlength) one need to multiply cfg.unitinmm with ppath to convert it to mm.
        % % output_det.p = detphoton.p; % exit position when cfg.issaveexit=1
        % % output_det.v = detphoton.v; % exit direction, when cfg.issaveexit=1
        % % output_det.prop = detphoton.prop;
        output_det = process_simulations(squeeze(Optical_prop(l,:,:)),info_model.cfg);

        % save output
        disp('Save results')
        if(s == 1)
            writematrix(output_det.ppath,strcat(out_path,'ppath_',num2str(Lambdas(l)),'.txt'),'Delimiter',' ');
            writematrix(output_det.p,strcat(out_path,'p_',num2str(Lambdas(l)),'.txt'),'Delimiter',' ');
            writematrix(output_det.v,strcat(out_path,'v_',num2str(Lambdas(l)),'.txt'),'Delimiter',' ');
            writematrix(output_det.prop,strcat(out_path,'prop_',num2str(Lambdas(l)),'.txt'),'Delimiter',' ');
        else
            writematrix(output_det.ppath,strcat(out_path,'ppath_',num2str(Lambdas(l)),'.txt'),'Delimiter',' ','WriteMode','append');
            writematrix(output_det.p,strcat(out_path,'p_',num2str(Lambdas(l)),'.txt'),'Delimiter',' ','WriteMode','append');
            writematrix(output_det.v,strcat(out_path,'v_',num2str(Lambdas(l)),'.txt'),'Delimiter',' ','WriteMode','append');
        end
        clear output_det;
    end
    
    % Zip files
    disp('Zip results')
    cd(out_path);
    zip(strcat(num2str(Lambdas(l)),'.zip'),{ ...
    %strcat('nscat_',num2str(Lambdas(l)),'.txt'), ...
    strcat('ppath_',num2str(Lambdas(l)),'.txt'), ...
    strcat('p_',num2str(Lambdas(l)),'.txt'), ...
    strcat('v_',num2str(Lambdas(l)),'.txt'), ...
    strcat('prop_',num2str(Lambdas(l)),'.txt')});

    % remove txt files
    disp('Delete temp files')
    % delete(strcat('nscat_',num2str(Lambdas(l)),'.txt'));
    delete(strcat('ppath_',num2str(Lambdas(l)),'.txt'));
    delete(strcat('p_',num2str(Lambdas(l)),'.txt'));
    delete(strcat('v_',num2str(Lambdas(l)),'.txt'));
    delete(strcat('prop_',num2str(Lambdas(l)),'.txt'));

    cd ..
end

File: Python WMC_simulations/functions/run_simulations.py
# -*- coding: utf-8 -*-
"""
Created on Mon Oct 28 12:18:08 2024

@author: angel
"""

import os
import numpy as np
import shutil
import pmcx
from process_model_info import process_model_info
from process_optical_properties import compute_optical_properties
from process_simulations import process_simulations
import matplotlib.pyplot as plt
# Divide the size of the pixel (increase the resolution)
division_factor = 1

# Model rectangle blood vessel
model_rect_blood_vessel = 0  # if 0, a pyramidal blood vessel is computed

# Inverse volume for display
inverse_volume_for_display = 0

Lambdas = np.arange(900, 910, 10)
run_in_cluster = 0
nb_repeat = 1  # Number of repetitions used in MCX
simu_repeat = 1  # Larger number of repeats (avoid large txt files)
nb_photons = 1e6

out_path = 'output/'
in_img_path = '../images/Synthetic_img/'

# Add paths
if run_in_cluster == 1:
    os.sys.path.append('/pbs/home/c/ccaredda/private/mcx/utils')
    os.sys.path.append('/pbs/home/c/ccaredda/private/mcxlab')
else:
    pass  # Add local paths if needed

# Create output directory
if not os.path.isdir(out_path):
    os.makedirs(out_path)

# Process model info
info_model = process_model_info(nb_photons, nb_repeat, "Simple shape", division_factor, model_rect_blood_vessel, inverse_volume_for_display)

# Save model info
np.save(os.path.join(out_path, 'cst.npy'), info_model)
# with open(os.path.join(out_path, 'cst.txt'), 'w') as f:
#     f.write(f'nb_photons {nb_photons}\n')
#     f.write(f'repetitions {nb_repeat * simu_repeat}\n')
#     f.write(f'unitinmm {info_model["unitinmm"]}\n')
#     f.write(f'vol_rows {info_model.cfg.vol.shape[0]}\n')
#     f.write(f'vol_cols {info_model.cfg.vol.shape[1]}\n')
#     f.write(f'division_factor {division_factor}\n')
#     f.write(f'model_rect_blood_vessel {model_rect_blood_vessel}\n')

# Process optical properties
Optical_prop = compute_optical_properties(Lambdas, 0)

# Process simulations
for l in range(len(Lambdas)):
    print(f"Simulation lambda {Lambdas[l]}")
    
    for s in range(simu_repeat):
        output_det = process_simulations(Optical_prop[l, :, :], info_model)
        np.squeeze(output_det['flux'])
        
        plt.imshow(np.log10(output_det['flux'][30,:, :]))
        plt.colorbar()
        plt.show()


        # Save output
    #     print('Save results')
    #     if s == 1:
    #         np.savetxt(os.path.join(out_path, f'ppath_{Lambdas[l]}.txt'), output_det.ppath, delimiter=' ')
    #         np.savetxt(os.path.join(out_path, f'p_{Lambdas[l]}.txt'), output_det.p, delimiter=' ')
    #         np.savetxt(os.path.join(out_path, f'v_{Lambdas[l]}.txt'), output_det.v, delimiter=' ')
    #         np.savetxt(os.path.join(out_path, f'prop_{Lambdas[l]}.txt'), output_det.prop, delimiter=' ')
    #     else:
    #         with open(os.path.join(out_path, f'ppath_{Lambdas[l]}.txt'), 'ab') as f:
    #             np.savetxt(f, output_det.ppath, delimiter=' ')
    #         with open(os.path.join(out_path, f'p_{Lambdas[l]}.txt'), 'ab') as f:
    #             np.savetxt(f, output_det.p, delimiter=' ')
    #         with open(os.path.join(out_path, f'v_{Lambdas[l]}.txt'), 'ab') as f:
    #             np.savetxt(f, output_det.v, delimiter=' ')

    #     del output_det

    # # Zip files
    # print('Zip results')
    # shutil.make_archive(os.path.join(out_path, str(Lambdas[l])), 'zip', out_path, base_dir=None, verbose=0, dry_run=False, owner=None, group=None, logger=None)

    # # Remove txt files
    # print('Delete temp files')
    # os.remove(os.path.join(out_path, f'ppath_{Lambdas[l]}.txt'))
    # os.remove(os.path.join(out_path, f'p_{Lambdas[l]}.txt'))
    # os.remove(os.path.join(out_path, f'v_{Lambdas[l]}.txt'))
    # os.remove(os.path.join(out_path, f'prop_{Lambdas[l]}.txt'))
